File: desktop/src/types/error-details.ts
/**
 * Detailed error information for rich error reporting
 */
export interface ErrorDetails {
  /** Error code for categorization (e.g., "context_length_exceeded") */
  code: string;
  /** User-friendly error message */
  message: string;
  /** Provider-specific error information if available */
  providerError?: ProviderErrorInfo;
  /** Whether a fallback to another provider was attempted */
  fallbackAttempted: boolean;
}

/**
 * Provider-specific error information
 */
export interface ProviderErrorInfo {
  /** Provider name (e.g., "openai", "google", "anthropic") */
  provider: string;
  /** HTTP status code from the provider */
  statusCode: number;
  /** Provider's error type (e.g., "invalid_request_error") */
  errorType: string;
  /** Full error details from the provider */
  details: string;
  /** Additional context about the error */
  context?: ErrorContext;
}

/**
 * Additional context for specific error types
 */
export interface ErrorContext {
  /** Number of tokens requested */
  requestedTokens?: number;
  /** Maximum tokens allowed by the model */
  maxTokens?: number;
  /** Model's context length limit */
  modelLimit?: number;
  /** Any other relevant context */
  additionalInfo?: string;
}

File: desktop/src-tauri/src/models/error_details.rs
use serde::{Deserialize, Serialize};

/// Detailed error information for rich error reporting
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ErrorDetails {
    /// Error code for categorization (e.g., "context_length_exceeded")
    pub code: String,
    /// User-friendly error message
    pub message: String,
    /// Provider-specific error information if available
    pub provider_error: Option<ProviderErrorInfo>,
    /// Whether a fallback to another provider was attempted
    pub fallback_attempted: bool,
}

/// Provider-specific error information
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderErrorInfo {
    /// Provider name (e.g., "openai", "google", "anthropic")
    pub provider: String,
    /// HTTP status code from the provider
    pub status_code: u16,
    /// Provider's error type (e.g., "invalid_request_error")
    pub error_type: String,
    /// Full error details from the provider
    pub details: String,
    /// Additional context about the error
    pub context: Option<ErrorContext>,
}

/// Additional context for specific error types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ErrorContext {
    /// Number of tokens requested
    pub requested_tokens: Option<i32>,
    /// Maximum tokens allowed by the model
    pub max_tokens: Option<i32>,
    /// Model's context length limit
    pub model_limit: Option<i32>,
    /// Any other relevant context
    pub additional_info: Option<String>,
}